modules/channel-teams/src/backend/client.ts

```diff
@@ -0,0 +1,181 @@
+import { BotFrameworkAdapter, ConversationReference, CardFactory, TurnContext, AttachmentLayoutTypes } from 'botbuilder'
+
+import * as sdk from 'botpress/sdk'
+import { Router } from 'express'
+import _ from 'lodash'
+
+import { Config } from '../config'
+
+import { Clients } from './typings'
+
+const outgoingTypes = ['message', 'typing', 'carousel', 'text']
+
+export class TeamsClient {
+  private conversationsRefs: _.Dictionary<Partial<ConversationReference>> = {}
+  private adapter: BotFrameworkAdapter
+  constructor(private bp: typeof sdk, private botId: string, private config: Config, private router: Router) {}
+
+  async initialize() {
+    this.adapter = new BotFrameworkAdapter({
+      appId: this.config.microsoftAppId,
+      appPassword: this.config.microsoftAppPassword
+    })
+
+    this.router.post('/api/messages', async (req, res) => {
+      await this.adapter.processActivity(req, res, async turnContext => {
+        const conversationReference = TurnContext.getConversationReference(turnContext.activity)
+        if (!turnContext.activity.text) {
+          // To prevent from emojis reactions to launch actual events
+          return
+        }
+
+        const threadId = conversationReference.conversation.id
+        this.conversationsRefs[threadId] = conversationReference
+        this._sendIncomingEvent(turnContext, threadId, { type: turnContext.activity.type })
+      })
+    })
+  }
+
+  private _sendIncomingEvent = (ctx: TurnContext, threadId: string, args: { type: string }) => {
+    this.bp.events.sendEvent(
+      this.bp.IO.Event({
+        botId: this.botId,
+        channel: 'teams',
+        direction: 'incoming',
+        payload: { text: ctx.activity.text },
+        preview: ctx.activity.text,
+        threadId,
+        target: ctx.activity.from.id,
+        ...args
+      })
+    )
+  }
+
+  public async sendOutgoingEvent(event: sdk.IO.Event): Promise<void> {
+    const messageType = event.type === 'default' ? 'text' : event.type
+
+    if (!_.includes(outgoingTypes, messageType)) {
+      throw new Error('Unsupported event type: ' + event.type)
+    }
+
+    const ref = this.conversationsRefs[event.threadId]
+
+    let msg: any = event.payload // TODO: place this logic in builtin with content-types
+    if (msg.type === 'typing') {
+      msg = {
+        type: 'typing'
+      }
+    } else if (msg.type === 'carousel') {
+      msg = this._sendCarousel(event)
+    } else if (msg.quick_replies && msg.quick_replies.length) {
+      msg = this._sendChoice(event)
```

Comment:
Can you rename those methods so they reflect what they actually do? They're not sending stuff, they are only preparing it for the send method

Reply:
> The documentation on how to set it up is missing details. Can you write all the steps necessary to set it up from scratch? Users shouldn't have to go through external documentation to do that.
> 
> Also, I couldn't figure out where api/messages is used or why it's there. Are we supposed to specify it somewhere ?

You're right, I have not specified where to use the 'api/message' endpoint. I will add some more information.